homeassistant/components/powerwall/__init__.py

```diff
@@ -87,26 +93,54 @@ def _async_migrator(entity_entry: entity_registry.RegistryEntry):
     await entity_registry.async_migrate_entries(hass, entry_id, _async_migrator)
 
 
+async def handle_setup_error(
+    hass: HomeAssistant, entry: ConfigEntry, error: PowerwallError
+):
+    """Log error and send notification for an error during the component setup."""
+    # Only emit one notification and log the error once
+    if hass.data[DOMAIN][entry.entry_id][POWERWALL_ERROR] is None:
+        # Only log the error message, the full traceback is not needed
+        _LOGGER.error(str(error))
+
+        if isinstance(error, PowerwallUnreachableError):
+            msg = "Could not connect to powerwall.\nSee logs for more information."
+            title = "Powerwall unreachable"
+        elif isinstance(error, APIChangedError):
+            msg = "It seems like the API of the powerwall changed.\nPlease report this issue so it can be resolved!\nSee logs for more information."
+            title = "Powerwall API changed"
+        else:
+            msg = "There was an error setting up the powerwall integration.\nSee logs for more information."
+            title = "Powerwall setup error"
+
+        hass.components.persistent_notification.async_create(msg, title=title)
+
+        hass.data[DOMAIN][entry.entry_id][POWERWALL_ERROR] = error
+
+
 async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry):
     """Set up Tesla Powerwall from a config entry."""
 
     entry_id = entry.entry_id
 
-    hass.data[DOMAIN].setdefault(entry_id, {})
+    hass.data[DOMAIN].setdefault(entry_id, {POWERWALL_ERROR: None})
     http_session = requests.Session()
     power_wall = Powerwall(entry.data[CONF_IP_ADDRESS], http_session=http_session)
     try:
         await hass.async_add_executor_job(power_wall.detect_and_pin_version)
         powerwall_data = await hass.async_add_executor_job(call_base_info, power_wall)
-    except (PowerwallUnreachableError, APIError, ConnectionError):
+    except PowerwallError as err:
         http_session.close()
+        await handle_setup_error(hass, entry, err)
         raise ConfigEntryNotReady
```

Comment:
@jrester The system is going to retry forever when `ConfigEntryNotReady` is raised.  We should only raise this if we expect the situation will resolve such as a transient network issue.  

I'm thinking `PowerwallUnreachableError` and `ConnectionError` get `ConfigEntryNotReady` and `APIError` gets the new code?

Reply:
Inside `handle_setup_error` it is first checked if we already had an error and only if it is new the notification is sent. This is then reset after the component was successfully setup. But I can also remove the notification part. Was quite unsure if it would be useful to inform the user about such failures.

But it might be sensible to completely refuse setup when an APIChangedError is received, because this means the API cannot be queried anymore.

Reply:
Is there any way to terminate the `DataUpdateCoordinator` from the `async_update_data` method? Because if we receive an `APIChangedError` we should completely stop polling.

Reply:
@jrester AFAICT `DataUpdateCoordinator` doesn't contemplate that scenario.  We could set a flag `hass.data[DOMAIN][entry.entry_id][UPDATE_PERM_FAILED] = True` (or check the error) and have `async_update_data` return when its set so the I/O is avoided.  Its not as clean as I'd like but is also like you said, an edge case.

Reply:
@bdraco The problem is that AFAIC we need to return correctly structured data otherwise the log gets filled with `TypeError`s because of the sensors trying to display it. I would return cached data but this could still fill the logs when it is the first request that is performed. The easiest workaround could be to request all data during the startup. But I am quite unsure if it is really worth the effort.

Reply:
@jrester Maybe something like this? https://github.com/bdraco/home-assistant/commit/adb659ba2b7ba22915daa4d6306a80d9ccd1a276

I haven't see where the `TypeError`s are coming from so I may be missing the issue though.

Reply:
@bdraco I had a similar idea but when the `async_update_data` returns None this gets set as data for the `DataUpdateCoordinator` and when the senor tries to use the data for example here `self._coordinator.data[POWERWALL_API_CHARGE]` this raises an error because `self._coordinator.data` is None and as such isn't subscriptable.

Reply:
If we `raise UpdateFailed("Powerwall api has changed")` instead of returning `None` does that work around the issue?